src/passes/SimplifyGlobals.cpp

```diff
@@ -250,21 +320,79 @@ struct SimplifyGlobals : public Pass {
     }
   }
 
-  void removeWritesToUnreadGlobals() {
-    // Globals that are not exports and not read from can be eliminated
-    // (even if they are written to).
-    NameSet unreadGlobals;
+  // Removes writes from globals that will never do anything useful with the
+  // written value anyhow. Returns whether an addition iteration is necessary.
+  bool removeUnneededWrites() {
+    bool more = false;
+
+    // Globals that are not exports and not read from are unnecessary (even if
+    // they are written to). Likewise, globals that are only read from in order
+    // to write to themselves are unnecessary. First, find such globals.
+    NameSet unnecessaryGlobals;
     for (auto& global : module->globals) {
       auto& info = map[global->name];
-      if (!info.imported && !info.exported && !info.read) {
-        unreadGlobals.insert(global->name);
+
+      if (!info.written) {
+        // No writes occur here, so there is nothing for us to remove.
+        continue;
+      }
+
+      if (info.imported || info.exported) {
+        // If the global is observable from the outside, we can't do anythng
+        // here.
+        //
+        // TODO: optimize the case of an imported but immutable global, etc.
+        continue;
+      }
+
+      // We only ever read-only-to-write if all of our reads are done in places
+      // we identified as read-only-to-write. That is, we have eliminated the
+      // possibility of any other uses. (Technically, each read-to-write
+      // location might have more than one read since we did not count them, but
+      // only verified there was one read or more; but this is good enough as
+      // the common case has exactly one.)
+      //
+      // Note that there might be more writes, if there are additional writes
+      // besides those in the read-only-to-write locations. But we can ignore
+      // those, as whatever they write will not be read in order to do anything
+      // of value.
+      bool onlyReadOnlyToWrite = (info.read == info.readOnlyToWrite);
+      assert(!onlyReadOnlyToWrite || info.written >= info.readOnlyToWrite);
```

Comment:
Not sure if I understand the purpose of this `assert`... Isn't `info.written >= info.readOnlyToWrite` always true, no matter `onlyReadOnlyToWrite` > 0 or not, because `written` is a superset of `readOnlyToWrite`?

Reply:
Good point, yeah, I'll fix that.